dvc/remote/local.py

```diff
@@ -739,3 +738,14 @@ def _log_missing_caches(self, checksum_info_dict):
                 "nor on remote. Missing cache files: {}".format(missing_desc)
             )
             logger.warning(msg)
+
+    def get_files_number(self, md5):
+        cache = self.get(md5)
+
+        if not cache:
+            raise CacheNotFoundException(md5)
+
+        if self.is_dir_cache(cache):
+            return len(self.load_dir_cache(md5))
```

Comment:
Loading dir cache for a giant directory might be quite expensive. How about we just update the n_total as we go? I.e. we are loading dir cache in do_checkout anyway, so why not just add the length of what we've discovered there along the way? What do you think?

Reply:
Hmm, I think it will degrade user experience when checking out, for example, two huge directories. However, I guess its better to have "self extending" progress indicator than waiting extended amount of time just to know how many files we are actually checking out.

Reply:
@pared Great point! Another advantage of "self-extending" approach, is that you'll be able to drop a lot of code (e.g. for calculating the total length and stuff), but yeah, self-extension is indeed misleading when it comes to actually keeping the user up to date. You could also just load dir cache once and use it where needed within your current approach.

Reply:
@efiop I like the idea of preloading cache, however I think it would be problematic. What possible way of implementation I see here, is to store caches as some `dict` in `Stage`, however then we would probably have to validate whether this `dict` is up to date whenever we would like to use it, right?

Reply:
@pared Yeah, I guess. E.g. storing it in Output along with the checksum, to detect if we need to reload it.

Reply:
@efiop Ok, so how about this: Ill implement "self extending" one as it is some kind of solution for current issue and make a new one that will focus on reorganizing loading mechanism?

Reply:
@pared Do you really feel like scrapping everything and re-writing it with "self extending" approach that you don't even like? :slightly_smiling_face: The dir_cache caching part is really simple, you just need to implement out.get_dir_cache() that would either load dir cache or return the one that is already loaded if self.checksum matches.

Reply:
@efiop Fair point, didn't realize that dir_cache caching is not that big.